TLS Verify Peers Fix

akutz opened the pull request:
This patch fixes the logic problem with verifyPeers. If verifyPeers is enabled and no peer is specified or there are no known_hosts files an error will now occur instead of falling through to a successful connection.

Many thanks to @vladimirvivien for discovering this issue.

vladimirvivien commented on drivers/storage/libstorage/libstorage_driver_tls.go:
```diff
@@ -45,49 +45,49 @@ func verifyKnownHost(
+	return false, newErrKnownHost(peerCerts)
 }
 
 func verifyKnownHostFiles(
```
I would change those two names, `verifyKnwonHostFiles` and `verifyKnownHostsFile`, to be more differentiated or document clearly how each is different.

akutz replied:
Hi @vladimirvivien,

Thank you for the suggestion. The patch doesn't include the addition of these functions as they're not new. They're also private, so I think it's not as important to name them completely different. Their purpose is pretty obvious since the first one is only 22 lines long and calls the second twice, once for each known_hosts file. I'm going to leave them as is for now for these reasons.

vladimirvivien reviewed (commented):
Looks good, minus the naming of these two funcs.

akutz committed:
TLS Verify Peers Fix

This patch fixes the logic problem with verifyPeers. If verifyPeers is
enabled and no peer is specified or there are no known_hosts files an
error will now occur instead of falling through to a successful
connection.

vladimirvivien commented:
@akutz 
This patch produces an error when `libstorage.server.tls:true` 
```
rexray:
  modules:
    default-docker:
      disabled: true
libstorage:
  embedded: true
  client:
    type: controller
  server:
    tls: true
    endpoints:
      public:
        address: tcp://:7979
    services:
      virtualbox:
        driver: virtualbox
```
Causes the following error when the RR service is starting.
```
Apr 24 17:54:57 ubuntu rexray[8935]: REX-Ray
Apr 24 17:54:57 ubuntu rexray[8935]: -------
Apr 24 17:54:57 ubuntu rexray[8935]: Binary: /home/vladimir/DEV/go/bin/rexray
Apr 24 17:54:57 ubuntu rexray[8935]: Flavor: client+agent+controller
Apr 24 17:54:57 ubuntu rexray[8935]: SemVer: 0.8.2+17+dirty
Apr 24 17:54:57 ubuntu rexray[8935]: OsArch: Linux-x86_64
Apr 24 17:54:57 ubuntu rexray[8935]: Branch: tls-support-583
Apr 24 17:54:57 ubuntu rexray[8935]: Commit: efda90972215aeeb5b1a5e5d557bf6296f6a070c
Apr 24 17:54:57 ubuntu rexray[8935]: Formed: Mon, 24 Apr 2017 17:38:15 EDT
Apr 24 17:54:57 ubuntu rexray[8935]: libStorage
Apr 24 17:54:57 ubuntu rexray[8935]: ----------
Apr 24 17:54:57 ubuntu rexray[8935]: SemVer: 0.5.2+38
Apr 24 17:54:57 ubuntu rexray[8935]: OsArch: Linux-x86_64
Apr 24 17:54:57 ubuntu rexray[8935]: Branch: 8570256
Apr 24 17:54:57 ubuntu rexray[8935]: Commit: 857025622f33431b47f072024680ec89c88d3982
Apr 24 17:54:57 ubuntu rexray[8935]: Formed: Mon, 24 Apr 2017 17:38:03 EDT
Apr 24 17:54:57 ubuntu rexray[8935]: time="2017-04-24T17:54:57-04:00" level=error msg="error starting libStorage server" error="tls: neither Certificates nor GetCertificate set in Config" time=1493070897036
Apr 24 17:54:57 ubuntu rexray[8935]: time="2017-04-24T17:54:57-04:00" level=error msg="default module(s) failed to initialize" error="tls: neither Certificates nor GetCertificate set in Config" time=1493070897037
Apr 24 17:54:57 ubuntu rexray[8935]: time="2017-04-24T17:54:57-04:00" level=error msg="daemon failed to initialize" error="tls: neither Certificates nor GetCertificate set in Config" time=1493070897037
Apr 24 17:54:57 ubuntu rexray[8935]: time="2017-04-24T17:54:57-04:00" level=error msg="error starting rex-ray" error="tls: neither Certificates nor GetCertificate set in Config" time=1493070897037
```
But, the error goes away when I completely remove `tls:true` setting.

akutz commented:
Hi @vladimirvivien,

That's because you told it to enable TLS on the server where there are no certificate files specified and none were found in the default locations. The server cannot enable TLS without certificates.

vladimirvivien commented:
@akutz the cert/key files were at the default `/etc/libstorage/tls` directory actually.  That's why I was surprised it did not work.

akutz commented:
Hi @vladimirvivien,

Not according to the provided logs they weren't (or they were -- you didn't enable DEBUG so I can't be sure). And you should remove it anyway to test the auto-enable. Please provide the full config and debug logs via a Gist link. Please indicate your deployment strategy as well.

akutz commented:
Hi @vladimirvivien,

According to the above log you're not using the TLS patch:

> `Apr 24 17:54:57 ubuntu rexray[8935]: Commit: 857025622f33431b47f072024680ec89c88d3982`

The TLS patch is ded0789bf209bcf390ef19c4c4d6b1156b8d9eab. You're still using 857025622f33431b47f072024680ec89c88d3982, the path refactor.

This is why the files didn't load. Please fetch libStorage `HEAD` and try again.
